ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,274 @@
 typedef struct _PNP_NOTIFY_ENTRY
 {
     LIST_ENTRY PnpNotifyList;
-    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
     PVOID Context;
-    UNICODE_STRING Guid;
-    PFILE_OBJECT FileObject;
     PDRIVER_OBJECT DriverObject;
     PDRIVER_NOTIFICATION_CALLBACK_ROUTINE PnpNotificationProc;
+    union
+    {
+        GUID Guid; // for EventCategoryDeviceInterfaceChange
+        struct // for EventCategoryTargetDeviceChange
+        {
+            PFILE_OBJECT FileObject;
+        };
+    };
+    IO_NOTIFICATION_EVENT_CATEGORY EventCategory;
+    volatile LONG Deleted;
 } PNP_NOTIFY_ENTRY, *PPNP_NOTIFY_ENTRY;
 
-KGUARDED_MUTEX PnpNotifyListLock;
-LIST_ENTRY PnpNotifyListHead;
+KGUARDED_MUTEX IopTargetDeviceNotifyLock;
+KGUARDED_MUTEX IopDeviceClassNotifyLock;
+KGUARDED_MUTEX IopHwProfileNotifyLock;
+
+LIST_ENTRY PiNotifyHwProfileListHead;
+LIST_ENTRY PiNotifyDeviceInterfaceListHead;
 
 /* FUNCTIONS *****************************************************************/
 
+CODE_SEG("INIT")
 VOID
-IopNotifyPlugPlayNotification(
-    IN PDEVICE_OBJECT DeviceObject,
-    IN IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
-    IN LPCGUID Event,
-    IN PVOID EventCategoryData1,
-    IN PVOID EventCategoryData2)
+PiInitializeNotifications(VOID)
 {
-    PPNP_NOTIFY_ENTRY ChangeEntry;
-    PLIST_ENTRY ListEntry;
-    PVOID NotificationStructure;
-    BOOLEAN CallCurrentEntry;
-    UNICODE_STRING GuidString;
-    NTSTATUS Status;
-    PDEVICE_OBJECT EntryDeviceObject = NULL;
+    KeInitializeGuardedMutex(&IopTargetDeviceNotifyLock);
+    KeInitializeGuardedMutex(&IopDeviceClassNotifyLock);
+    KeInitializeGuardedMutex(&IopHwProfileNotifyLock);
+    InitializeListHead(&PiNotifyHwProfileListHead);
+    InitializeListHead(&PiNotifyDeviceInterfaceListHead);
+}
 
-    ASSERT(DeviceObject);
+CODE_SEG("PAGE")
+VOID
+PiFreeTargetDeviceNotifyList(
+    _In_ PDEVICE_NODE DeviceNode)
+{
+    PAGED_CODE();
 
-    KeAcquireGuardedMutex(&PnpNotifyListLock);
-    if (IsListEmpty(&PnpNotifyListHead))
+    for (PLIST_ENTRY entry = DeviceNode->TargetDeviceNotify.Flink;
```

Comment:
Declaration should be on top the function / scope.

Reply:
@tkreuzer: Frankly asking, what are the benefits?
As far as I know, C89 requiring this is no "feature", but a limitation of the compilers of that time.
However, our currently used compilers are no longer bounded by the artificial limits of C89.

I only see downsides here: More lines of code and cluttered scope definitions. A perfect example is: https://github.com/reactos/reactos/blob/30144179190c11519add8c1f4fe13afed52cb997/ntoskrnl/io/iomgr/file.c#L324-L361